qip-0011.md

```diff
@@ -0,0 +1,128 @@
+speculators buy up too many account slots to resell them to future buyers, they
+run the risk of increasing block processing latency, which through
+[QIP-0008](qip-0008.md) will ultimately lead to the addition of more chains
+with more resources. If this happens, speculators will have to compete with
+these newly available resources, which would harm their speculative investment.
+So there is some negative feedback here, which incentivizes speculators to
+participate and help determine market pricing, without fully consuming
+resources that would trigger a trie expansion. In fact, when the network gets
+close to that point, there is a strong incentive for account holders to sell
+back any account space they no longer need, which helps balance resource usage
+as the network scales.
+
+## Specification
+### Account Pricing
+Every new account added to the tree must pay for its address space. The cost of
+this address space must increase as the number of addresses increases. This
+will create negative feedback: speculators will occupy address space with new
+accounts when its cheap, and destroy their accounts when if it becomes
+expensive. This balance of speculator preference with user demand, is how the
+market determines the appropriate address space price.
+
+To achieve this, we use the radix-16 computational bound described above, to
+set the account slot price. The price function is:
+
+$$ P_a = K_a* \lfloor log_{16}(N) \rfloor $$
+
+where $P_a$ is the price to create or sell an account slot in Quai per account,
+$N$ is the total number of accounts in the trie, and $K_a$ is a constant
+scaling factor chosen to adjust price responsiveness to trie size.
+
+#### Choosing $K_a$
+We choose $K_a$ to set a Schelling point about a reasonable acceptable trie
+size, while understanding that the actual size will again depend on the
+subjective demand for account space. We choose 1 billion (2^30) accounts per
+chain as a reasonable upper bound, and we choose 0.1 Quai per account to be a
+high cost which will lead to reduced demand in the account space market. 
+
+We choose:
+
+$$ K_a = 0.1*\frac{Quai}{account^2} $$
+
+$$ \approx 9.31*10^{-11} \frac{Quai}{account^2} $$
+
+### Protocols for Buying and Selling Account Slots
+ For a new account to be created, enough Quai must be sent to cover the account
+ creation price $P_a$. The new account will be credited with the balance of the
+ transaction minus the creation price, $P_a$. Any transaction to a new account,
+ which does not satisfy the creation price, will fail.
+
+Conversely, to sell account space back, we provide a precompiled contract with
+a destruction method. The destruction method will allow the user to provide a
+"beneficiary transaction", which may transfer the balance including the sale
+price of the account (along with optional TX data) to any other address in the
+trie. Upon successful execution of the transaction, the account will be deleted
+from the state trie.
```

Comment:
This doesn't seem like a sale to me. If the market price of an account in the state trie increases, I should be able to sell my account to the highest bidder. This would involve the transfer of my account space to them, not the destruction of my account space.

Reply:
Another question: a smart contract (usually) takes up more space than an account. Should a smart contract pay more for the bytes it stores?

Reply:
You are selling it to the protocol, but market demand dictates the price. Think of it like an account space LP.

The first person to buy an account will get it basically for free. If the network grows to have more users buying up account space, then the trie will get bigger and the account price will go up. That first person can then sell back his account to the protocol at a higher price, which is reflective of the increased demand for account space since he bought his account.

Similarly, someone who buys an account when the trie is very large, might not get much money back if other users sell their accounts and shrink the trie size.

Reply:
This will not create a market. If the market price is high or low relative to the protocol price, either the buyer or seller will always do better by taking the protocol price. Ergo, there is no market because the protocol price always guarantees that one of the two participants will use the protocol, thus eliminating the market.

Reply:
You're right that this isn't a market, and instead its a demand curve (logarithmic in this case) which adjusts price according to the demand for account space, and resource buys/sells are settled by the protocol.

To get a true market, you would need to enable direct buying and selling of account space. The idea I had for that was along the lines of this:
* add one more data field to each account leaf: a sell TX
* any user willing to sell their account should populate that field with the TX parameters for the sale (recipient, TX data, gas, value, etc)
* anyone who wishes to buy that account space must submit the sell TX from their account AND specify a new address to own that space

This has a few very challenging downsides:

1) You must already have an account to buy account space. So how are new accounts created? What if there are no sellers (i.e. in early days)? **This is what led me to the demand curve approach, because accounts must always be available to buy even when no sellers exist.**

2) I considered making a "startup window" early on, where new accounts can be created by the protocol. After that window new users must buy existing account space from sellers, but even that has impractical issues: what price do you sell the initial accounts at? This too led me to the demand curve approach, if only in a "startup window".

3) What about startup windows on new chains? Now you can end up in a scenario where some shards have wildly different account spaces available, which would imbalance the scaling properties of our hierarchy. I'm not sure how to reconcile that imbalance.

4) I considered making account space mined: the only way the protocol can create an account is by mining a block. This is the most tractable alternative, IMO, but still has the problem that new accounts may not be created fast enough to onboard a wave of new users. It could result in serious UX challenges. It would also require miners to generate new coinbase addresses for each block, which complicates the mining software.

5) Another alternative was to set an fixed account space limit in each chain and let there be a land rush to buy up all accounts when each chain starts. This prevents the market from determining the total account space, however, which could be problematic for several reasons. Pick an account space that is too small or large, and the protocol pays in under/over utilization. If underutilized, then you get in the way of dynamic expansion triggers. If over utilized, performance will suffer disproportionately to user demand.

All of the non demand-curve ideas above also have an undesirable scenario if account space becomes unavailable or incredibly scarce. What then? A new user is expected to pay 1M Quai to join the network?
 
The demand curve was the only way I saw to address all of these problems, not to mention its implementation is also the simplest (KISS). But yes... it is not a free market.

Reply:
tl;dr: its not a market, but I think a curve is the best we can do

Reply:
from Dr. K, demand curve is OK, but need to dynamically price load/store operations according to tree depth

Reply:
I've updated the spec to include dynamic gas cost for load/store operations. I believe this addresses all concerns?